### PR TITLE
Fireworks

### PR DESCRIPTION
**Pull request by Frago9876543210**
## Introduction
I think it's all clear from the title.

### API changes
For example:
```php
public function onClick(PlayerInteractEvent $e) : void{
	if($e->getAction() === $e::RIGHT_CLICK_BLOCK && $e->getItem()->isNull()){
		$item = new Fireworks;
		$item->setNamedTag(Fireworks::createNBT(new FireworksData(1, [
			new FireworksExplosion(Fireworks::COLOR_ORANGE, Fireworks::COLOR_CYAN, true, true, Fireworks::TYPE_STAR_SHAPED)
		])));
		$e->getPlayer()->getInventory()->addItem($item);
	}
}
```

image

### CODE REVIEW

File: src/pocketmine/item/Fireworks.php [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+		$rocket = new FireworksRocket($player->level, $nbt, $player, $this, $random);
+		if($player->isSurvival()){
+			--$this->count;
+		}
+		$rocket->spawnToAll();
+		return true;
+	}
+
+	public static function createNBT(FireworksData $fireworksData) : CompoundTag{
```
Comment (dktapps) [inline src/pocketmine/item/Fireworks.php]: This appears to be dead code.

File: src/pocketmine/item/FireworksData.php [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+ * @link http://www.pocketmine.net/
+ *
+ *
+*/
+
+declare(strict_types=1);
+
+namespace pocketmine\item;
+
+class FireworksData{
```
Comment (dktapps) [inline src/pocketmine/item/FireworksData.php]: This appears to be dead code.

File: src/pocketmine/item/FireworksExplosion.php [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+ * @link http://www.pocketmine.net/
+ *
+ *
+*/
+
+declare(strict_types=1);
+
+namespace pocketmine\item;
+
+class FireworksExplosion{
```
Comment (dktapps) [inline src/pocketmine/item/FireworksExplosion.php]: This appears to be dead code.

File: src/pocketmine/item/FireworksExplosion.php [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	 */
+	public function getFade() : string{
+		return $this->fade !== null ? chr($this->fade) : "";
+	}
+
+	/**
+	 * @return int
+	 */
+	public function isFlickering() : int{
+		return (int) $this->flicker;
```
Comment (dktapps) [inline src/pocketmine/item/FireworksExplosion.php]: This is very confusing. A plugin developer using this code will not understand why this doesn't return a boolean.

File: src/pocketmine/item/FireworksExplosion.php [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+	}
+
+	/**
+	 * @return int
+	 */
+	public function hasTrail() : int{
+		return (int) $this->trail;
```
Comment (dktapps) [inline src/pocketmine/item/FireworksExplosion.php]: Same as above.

File: src/pocketmine/entity/projectile/FireworksRocket.php [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+		parent::__construct($level, $nbt, $shootingEntity);
+	}
+
+	protected function initEntity(CompoundTag $nbt) : void{
+		parent::initEntity($nbt);
+		$random = $this->random ?? new Random;
+
+		$this->setGenericFlag(self::DATA_FLAG_HAS_COLLISION, true);
+		$this->setGenericFlag(self::DATA_FLAG_AFFECTED_BY_GRAVITY, true);
+		$this->getDataPropertyManager()->setItem(self::DATA_MINECART_DISPLAY_BLOCK, $this->item);
```
Comment (dktapps) [inline src/pocketmine/entity/projectile/FireworksRocket.php]: yes, this reads well indeed...

**Review by dktapps — COMMENTED** [review]
This appears to be incomplete. There are several dead code paths and the crafting capability is not present. As is, this would have taken me 5 minutes to do myself because the meat isn't here.

File: src/pocketmine/entity/projectile/FireworksRocket.php (hunk 6 above)
Comment (Frago9876543210) [reply src/pocketmine/entity/projectile/FireworksRocket.php]: wdym?

Comment (Muqsit) [reply src/pocketmine/entity/projectile/FireworksRocket.php]: code readability

**Comment by dktapps** [comment]
Some additional comments on why this was closed (aside from being stale) is that it is far too dependent on Mojang internals for me to feel comfortable touching it. We've seen how easily such things catch fire in the last year or so.

**Pull request closed without merging**